tools/topology/topology2/doc/topology2-generate-contents.sh

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2023 Intel Corporation. All rights reserved.
+# This script generates the contents page and stores it in extra-contents/contents.doxy
+
+set -e
+
+cd "$(dirname "$0")"
+cat > extra-contents/contents.doxy <<EOF
+/*! \page doc_contents
+    \section doc_classes Classes
+EOF
+
+for ctype in Base Widget Control Dai Pipeline
+do
+	echo "    \subsection doc_${ctype,}_classes $ctype Classes" >>  extra-contents/contents.doxy
+	find .. -name \*.conf | xargs grep -h ^Class\\.$ctype\\. | \
+		awk '{
+			split($1, a, "\\.");
```

Comment:
- `awk` has `grep` built-in.
- You don't need `split`, `-F` is enough
- use `-print0`

```shell

find .. -name \*.conf -print0 |
  xargs -0 awk -F. "/^Class\.${ctype}\./ "'{ name=$3; ... }'


```

Reply:
Otherwise Ok suggestions except one. Will follow the rest.

> * `awk` has `grep` built-in.
> * You don't need `split`, `-F` is enough

It is not. I need first white-space field separator. Then after I have separated the first field with white-space, then I need to split that with '.' as the separator, to pick the class name from:

Class.Base.<class_name> {

There are of course multiple different ways to extract <class_name> from a line like that, I have picked the first that came to my mind, and it works well enough.

> * use `-print0`
> * new lines are automatically escaped after a pipe `|` (or anywhere the shell expects more), so no need for a backslash there (and yes this is a very good point to insert a new line)
> 
> ```shell
> find .. -name \*.conf -print0 |
>   xargs -0 awk -F. "/^Class\.${ctype}\./ "'{ name=$3; ... }'
> ```